sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+      builder.addIfNotNull(DisplayData.item("filter", getFilter()));
+    }
+  }
+
+  /**
+   * A Hive {@link BoundedSource} reading {@link HCatRecord} from  a given instance.
+   */
+  @VisibleForTesting
+  static class BoundedHiveSource extends BoundedSource<HCatRecord> {
+    private Read spec;
+
+    private BoundedHiveSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public Coder<HCatRecord> getDefaultOutputCoder() {
+      return HCatRecordSimpleCoder.of(spec.getMetastoreUri());
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public BoundedReader<HCatRecord> createReader(PipelineOptions options) {
+      return new BoundedHiveReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws
+    IOException, NoSuchObjectException, TException {
+      //TODO: This is not really used, should be removed
+      Configuration conf = new Configuration();
+      conf.set(HCatConstants.HCAT_METASTORE_URI, spec.getMetastoreUri());
+      HiveConf hiveConf = HCatUtil.getHiveConf(conf);
+      IMetaStoreClient client = HCatUtil.getHiveMetastoreClient(hiveConf);
+      Table table = HCatUtil.getTable(client, spec.getDatabase(),
+        spec.getTable());
+      client.close();
+      return StatsUtils.getFileSizeForTable(hiveConf, table);
+    }
+
+    @Override
+    public List<BoundedSource<HCatRecord>> split(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      //TODO: We are really not using the bundle size at run time
+      //to make the split
+      //just going with the default splits returned by Hive
+      //unable to explicitly control the splits returned by Hive
+      List<BoundedSource<HCatRecord>> sources = new ArrayList<>();
+      ReaderContext cntxt = getReaderContext();
+      for (int i = 0; i < cntxt.numSplits(); i++) {
+        //create an instance of BoundedHiveSource for each of the split
+        //to be read in parallel
+        sources.add(new BoundedHiveSource(spec.withSplitId(i).withContext(cntxt)));
+      }
+      return sources;
+    }
+
+    private ReaderContext getReaderContext() {
+      ReadEntity entity = new ReadEntity.Builder().withDatabase(spec.getDatabase())
+          .withTable(spec.getTable()).withFilter(spec.getFilter()).build();
+      Map<String, String> hiveConfig = new HashMap<String, String>();
+      hiveConfig.put(HCatConstants.HCAT_METASTORE_URI, spec.getMetastoreUri());
+      HCatReader masterReader = DataTransferFactory.getHCatReader(entity, hiveConfig);
+      ReaderContext cntxt = null;
+      try {
+        cntxt = masterReader.prepareRead();
+        } catch (HCatException e) {
+        LOG.error("Exception while preparing for read", e);
```

Comment:
In this case you will return null - is this correct? Seems like a better action is to simply rethrow the exception (or to not catch it), because there is no way to recover from it.